plugin/pkg/admission/eventratelimit/admission.go

```diff
@@ -0,0 +1,270 @@
+type cache interface {
+	// get the rate limiter associated with the specified key
+	get(key interface{}) (rateLimiter flowcontrol.RateLimiter, found bool)
+	// add the specified rate limiter to the cache, associated with the specified key
+	add(key interface{}, rateLimiter flowcontrol.RateLimiter)
+}
+
+// singleCache is a cache that only stores a single, constant item
+type singleCache struct {
+	// the single rate limiter held by the cache
+	rateLimiter flowcontrol.RateLimiter
+}
+
+func newSingleCache(rateLimiter flowcontrol.RateLimiter) *singleCache {
+	return &singleCache{
+		rateLimiter: rateLimiter,
+	}
+}
+
+func (c *singleCache) get(key interface{}) (flowcontrol.RateLimiter, bool) {
+	return c.rateLimiter, true
+}
+
+func (c *singleCache) add(key interface{}, rateLimiter flowcontrol.RateLimiter) {
+	// Do nothing as the rate limiter stored by this cache is constant
+}
+
+// lruCache is a least-recently-used cache
+type lruCache struct {
+	cache *lru.Cache
+}
+
+func newLRUCache(cacheSize int) *lruCache {
+	return &lruCache{
+		cache: lru.New(cacheSize),
+	}
+}
+
+func (c *lruCache) get(key interface{}) (flowcontrol.RateLimiter, bool) {
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, false
+	}
+	return value.(flowcontrol.RateLimiter), true
+}
+
+func (c *lruCache) add(key interface{}, rateLimiter flowcontrol.RateLimiter) {
+	c.cache.Add(key, rateLimiter)
+}
+
+// factory for creating a rate limiter
+type rateLimiterFactory func(qps float32, burst int) flowcontrol.RateLimiter
+
+// newEventRateLimit configures an admission controller that can enforce event rate limits
+func newEventRateLimit(config *eventratelimitapi.Configuration) (admission.Interface, error) {
+	rlf := func(qps float32, burst int) flowcontrol.RateLimiter {
+		return flowcontrol.NewTokenBucketRateLimiter(qps, burst)
+	}
+	return newEventRateLimitUsingRLF(config, rlf)
+}
+
+// newEventRateLimitWithClock configures an admission controller that can enforce event rate limits.
+// It uses a clock for testing purposes.
+func newEventRateLimitWithClock(config *eventratelimitapi.Configuration, clock flowcontrol.Clock) (admission.Interface, error) {
+	rlf := func(qps float32, burst int) flowcontrol.RateLimiter {
+		return flowcontrol.NewTokenBucketRateLimiterWithClock(qps, burst, clock)
+	}
+	return newEventRateLimitUsingRLF(config, rlf)
+}
+
+// newEventRateLimit configures an admission controller than can enforce event rate limits
+func newEventRateLimitUsingRLF(config *eventratelimitapi.Configuration, rlf rateLimiterFactory) (admission.Interface, error) {
+	eventRateLimitAdmission := &eventRateLimitAdmission{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+
+	for _, limitConfig := range config.Limits {
+		enforcer, err := newLimitEnforcer(limitConfig, rlf)
+		if err != nil {
+			return nil, err
+		}
+		eventRateLimitAdmission.limitEnforcers = append(eventRateLimitAdmission.limitEnforcers, enforcer)
```

Comment:
nit: construct this array above and initialize the `eventRateLImitAdmission` in one init block instead of mutating it here.